baseapp/baseapp.go

```diff
@@ -568,6 +568,12 @@ func (app *BaseApp) getState(mode runTxMode) *state {
 	return app.checkState
 }
 
+// NewProcessProposalContext returns a context with a branched version of the
+// DeliverTx state that is safe to query during ProcessProposal.
+func (app *BaseApp) NewProcessProposalQueryContext() (sdk.Context, error) {
+	return app.createQueryContext(app.cms.LatestVersion(), false)
+}
```

Comment:
posting the `createQueryContext` logic here as its important, and afaiu, is doing exactly what we want. We need to ensure that we are not using the checkTx state, as that would not be deterministic, and we also want to make sure that we are using a branch of the latest state just to be safe.

https://github.com/celestiaorg/cosmos-sdk/blob/d4a340f4a10bcd0996c969303b072cc6f205e552/baseapp/abci.go#L622-L672